src/library_tty.js

```diff
@@ -63,13 +63,24 @@ mergeInto(LibraryManager.library, {
           if (result === null || result === undefined) break;
           bytesRead++;
           buffer[offset+i] = result;
+
+          // If EOL character, return having put it in the buffer.
+          if (result === 10) break;
         }
         if (bytesRead) {
           stream.node.timestamp = Date.now();
         }
         return bytesRead;
       },
       write: function(stream, buffer, offset, length, pos) {
+        if (Module['printChars'] != undefined) {
+          Module['printChars'](stream.fd, buffer.subarray(offset, offset + length), length);
```

Comment:
Instead of `buffer.subarray(offset, offset + length)`, which allocates garbage, I think it would be better to define the function to go for something like `Module['printChars'](stream.fd, buffer, offset, length);`?

Reply:
This is because as it stands UTF8ArrayToString is expecting to be passed a nul terminated string.  A subarray works as well because an out-of-bounds index ref returns undefined.  It might be more efficient to have an optional 'length' parameter on UTF8ArrayToString, maybe?

Unless we can assume we'll only be calling printChars with nul terminated strings in the first place, but then having a 'length' parameter implies that's not necessarily the case?

Reply:
Latest comment adds optional 'len' parameter to UTF8ArrayToString